src/OpenTelemetry.Api/Context/Propagation/BaseHeaderParser.cs

```diff
@@ -0,0 +1,83 @@
+﻿// <copyright file="BaseHeaderParser.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+namespace OpenTelemetry.Context.Propagation
+{
+    // Adoptation of code from https://github.com/aspnet/HttpAbstractions/blob/07d115400e4f8c7a66ba239f230805f03a14ee3d/src/Microsoft.Net.Http.Headers/BaseHeaderParser.cs
+    internal abstract class BaseHeaderParser<T> : HttpHeaderParser<T>
```

Comment:
I understand these are copied from the ASP.NET codebase, stepping back questions:
1. Do we need all the logic, or only a portion that we'll use for OpenTelemetry project?
2. Do we expect to keep the file "almost in-sync" with the original file in the future, or the plan is to refactor/cleanup?
3. What about test cases?

Reply:
Actually copied from AppInsights, which copied from ASP.NET 😄 See: 
https://github.com/aspnet/Microsoft.AspNet.TelemetryCorrelation/blob/e867e87e3a009223685b553dbb667b08bb9cab0c/src/Microsoft.AspNet.TelemetryCorrelation/ActivityExtensions.cs#L93 & https://github.com/aspnet/Microsoft.AspNet.TelemetryCorrelation/tree/master/src/Microsoft.AspNet.TelemetryCorrelation/Internal

I agree though, it's a lot of stuff. If we get general approval on the rest of the design, we can take a stab at simplifying it. All this is doing is parsing Baggage the from string(s) sent in header(s) into IEnumerable<KeyValuePair<string, string>>.

Reply:
I think the direction looks good, although the spec is still changing (as pointed out in the description of this PR) and we might need to change the implementation again.

I also like the fact that we've removed more lines than we added in this PR 👍

The borrowed code is good enough for this PR, we can simplify the flow and make it as "correct" as possible (based on the latest draft W3C spec) in another PR.